nixos/modules/services/hardware/nvidia-container-toolkit/default.nix

```
@@ -69,14 +69,16 @@
     virtualisation.docker.daemon.settings = lib.mkIf
       (config.hardware.nvidia-container-toolkit.enable &&
        (lib.versionAtLeast config.virtualisation.docker.package.version "25")) {
-      features.cdi = true;
-    };
+         features.cdi = true;
+       };
 
     hardware.nvidia-container-toolkit.mounts = let
       nvidia-driver = config.hardware.nvidia.package;
     in (lib.mkMerge [
       [{ hostPath = pkgs.addDriverRunpath.driverLink;
          containerPath = pkgs.addDriverRunpath.driverLink; }
+       { hostPath = "${lib.getLib nvidia-driver}/share";
+         containerPath = "${lib.getLib nvidia-driver}/share"; }
```

Might as well mount the full store path?

```suggestion

       { hostPath = "${lib.getLib nvidia-driver}";
         containerPath = "${lib.getLib nvidia-driver}"; }
```

---

The nvidia-container-toolkit cdi generator is cherry picking mounts inside `lib`. I would not override that by mounting `lib` itself.

---

I'm not sure I understand? What does the generator care about `/nix/store/...` paths?

---

It discovers libraries due to `--library-search-path` at https://github.com/NixOS/nixpkgs/blob/2ab53fad627302ae55a11548f85e5a653b42136d/nixos/modules/services/hardware/nvidia-container-toolkit/cdi-generate.nix#L29

---

Ah, I suppose that builds an LD_LIBRARY_PATH or an `ld.so.cache`.
Are you concerned nvidia-ctk might get confused about mounting both `${getLib nvidia-driver}/lib` and its parent `${getLib nvidia-driver}`?

We will be mounting complete store paths in the end, when we start handling the closure

---

Not really a concern, just a matter of avoiding redundancy. If we do that -- add the whole `lib.getLib nvidia-driver`, we can see on the generated `/var/run/cdi-nvidia-container-toolkit.json` file:

```json
{
  ...,
  "containerEdits": [
    "mounts": [
      ...,
      {
        "containerPath": "/nix/store/fdxrjmwy9v8qkw1bh9mnqhja16wnf4gr-nvidia-x11-550.78-6.6.32",
        "hostPath": "/nix/store/fdxrjmwy9v8qkw1bh9mnqhja16wnf4gr-nvidia-x11-550.78-6.6.32",
        "options": [
          "ro",
          "nosuid",
          "nodev",
          "bind"
        ]
      },
      ...
    ]
  ],
  ...
}
```
In addition to that, we are mounting cherry-picked libraries. Pasting just an example of one:

```
{
  ...,
  "containerEdits": [
    "mounts": [
      ...,
      {
        "hostPath": "/nix/store/fdxrjmwy9v8qkw1bh9mnqhja16wnf4gr-nvidia-x11-550.78-6.6.32/lib/libcuda.so.550.78",
        "containerPath": "/nix/store/fdxrjmwy9v8qkw1bh9mnqhja16wnf4gr-nvidia-x11-550.78-6.6.32/lib/libcuda.so.550.78",
        "options": [
          "ro",
          "nosuid",
          "nodev",
          "bind"
        ]
      },
      ...
    ]
  ],
  ...
}
```

So for me in this case, with the current `jq` patching, is more a matter of avoiding redundant mounts. It's not a big thing but I'd propose improving the whole situation when we move to mounting the whole closure with the script and `exportReferencesGraph`.

In the improved solution of using the `exportReferencesGraph`, I expect either:

1. The python script to filter out the mounts where `containerPath` starts with the nix closure path. Or,
2. Somehow tell the `nvidia-container-toolkit cdi generate` command to skip the libs. I don't know if we can make a case to contribute this upstream.